File: src/decoders/timetable-class-detention.ts
import type { SessionHandle, TimetableClassDetention } from "~/models";
import { decodeTimetableClass } from "./timetable-class";

export const decodeTimetableClassDetention = (item: any, session: SessionHandle): TimetableClassDetention => {
  let title: string | undefined;

  const personalNames: string[] = [];
  const teacherNames: string[] = [];
  const classrooms: string[] = [];

  if ("ListeContenus" in item) {
    for (const data of item.ListeContenus.V) {
      const label = data.L;

      if (data.estHoraire) {
        title = label;
      }
      else if ("G" in data) {
        switch (data.G) {
          case 3: // Teacher
            teacherNames.push(label);
            break;
          case 34: // Personal
            personalNames.push(label);
            break;
          case 17: // Room
            classrooms.push(label);
            break;
        }
      }
    }
  }

  return {
    is: "detention",
    ...decodeTimetableClass(item, session),

    title,
    personalNames,
    teacherNames,
    classrooms
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/timetable-class-detention.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.TimetableClassDetention
import kotlinx.serialization.json.*

fun decodeTimetableClassDetention(item: JsonObject): TimetableClassDetention {
  var title: String? = null

  val personalNames: MutableList<String> = mutableListOf()
  val teacherNames: MutableList<String> = mutableListOf()
  val classrooms: MutableList<String> = mutableListOf()

  if (item.containsKey("ListeContenus")) {
    for (data in item["ListeContenus"]!!.jsonObject["V"]!!.jsonArray) {
      val dataObj = data.jsonObject
      val label = dataObj["L"]!!.jsonPrimitive.content

      if (dataObj["estHoraire"]?.jsonPrimitive?.boolean == true) title = label
      else if (dataObj.containsKey("G"))
          when (dataObj["G"]!!.jsonPrimitive.int) {
            3 -> teacherNames.add(label)
            34 -> personalNames.add(label)
            17 -> classrooms.add(label)
          }
    }
  }

  return TimetableClassDetention(
      title = title,
      personalNames = personalNames,
      teacherNames = teacherNames,
      classrooms = classrooms)
}
